File: 2017/08/08.py
class Operation:
    def __init__(self, register, action, action_value, cond_register, cond_symbol, cond_value):
        self.register = register
        self.action = action
        self.action_value = action_value
        self.cond_register = cond_register
        self.cond_symbol = cond_symbol
        self.cond_value = cond_value

    def run(self, registers, all_time_max):
        # Check if condition is fullfilled
        condition_ok = False
        actual = registers[self.cond_register]
        expected = self.cond_value
        
        if self.cond_symbol == '==':
            if actual == expected:
                condition_ok = True
        elif self.cond_symbol == '!=':
            if actual != expected:
                condition_ok = True
        elif self.cond_symbol == '<':
            if actual < expected:
                condition_ok = True
        elif self.cond_symbol == '<=':
            if actual <= expected:
                condition_ok = True
        elif self.cond_symbol == '>':
            if actual > expected:
                condition_ok = True
        elif self.cond_symbol == '>=':
            if actual >= expected:
                condition_ok = True
        else:
            print("Unknown symbol: " + self.cond_symbol)
            raise 
        
        if not condition_ok:
            return all_time_max

        if self.action == 'inc':
            registers[self.register] += self.action_value
        else:
            registers[self.register] -= self.action_value

        if registers[self.register] > all_time_max:
            all_time_max = registers[self.register]
        return all_time_max
        
def parse_line_to_operation(line):
    parts = line.split()

    identifier = parts[0]
    action = parts[1]
    if action != 'inc' and action != 'dec':
        print('Wrong action: ' + action)
        return None
    number = int(parts[2])
    if parts[3] != 'if':
        print('Missing if: ' + parts[3])
        return None
    identifier2 = parts[4]
    symbol = parts[5]
    number2 = int(parts[6])

    return Operation(identifier, action, number, identifier2, symbol, number2)

with open('input', 'r') as f:
    lines = f.read().splitlines()

operations = [parse_line_to_operation(l) for l in lines]

# Initialize registers
registers = {}
for op in operations:
    if not op.register in registers:
        registers[op.register] = 0

# Run all operations
all_time_max = -2**32
for op in operations:
    all_time_max = op.run(registers, all_time_max)

# Display the largest value in the registers
m = 0
for value in registers.values():
    if value > m:
        m = value
print("Max: " + str(m))
print("All time max: " + str(all_time_max))

File: 2017/08/08.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

type Operation struct {
	register     string
	action       string
	actionValue  int
	condRegister string
	condSymbol   string
	condValue    int
}

func parseLine(line string) Operation {
	parts := strings.Fields(line)

	identifier := parts[0]
	action := parts[1]
	if action != "inc" && action != "dec" {
		panic("Wrong action: " + action)
	}
	number, err := strconv.Atoi(parts[2])
	if err != nil {
		panic(err)
	}
	if parts[3] != "if" {
		print("Missing if: " + parts[3])
	}
	identifier2 := parts[4]
	symbol := parts[5]
	number2, err := strconv.Atoi(parts[6])
	if err != nil {
		panic(err)
	}

	return Operation{identifier, action, number, identifier2, symbol, number2}
}

func (op Operation) run(registers map[string]int, allTimeMax int) int {
	// Check if condition is fullfilled
	var conditionOk bool

	fmt.Printf("---- Run op: %s %s %d if %s %s %d\n", op.register, op.action, op.actionValue, op.condRegister, op.condSymbol, op.condValue)
	fmt.Printf("Before: %s = %d\n", op.register, registers[op.register])

	actual := registers[op.condRegister]
	expected := op.condValue

	if op.condSymbol == "==" {
		if actual == expected {
			conditionOk = true
		}
	} else if op.condSymbol == "!=" {
		if actual != expected {
			conditionOk = true
		}
	} else if op.condSymbol == "<" {
		if actual < expected {
			conditionOk = true
		}
	} else if op.condSymbol == "<=" {
		if actual <= expected {
			conditionOk = true
		}
	} else if op.condSymbol == ">" {
		if actual > expected {
			conditionOk = true
		}
	} else if op.condSymbol == ">=" {
		if actual >= expected {
			conditionOk = true
		}
	} else {
		panic("Unknown symbol: " + op.condSymbol)
	}

	if !conditionOk {
		fmt.Println("DO NOT RUN.")
		return allTimeMax
	}

	if op.action == "inc" {
		registers[op.register] += op.actionValue
	} else {
		registers[op.register] -= op.actionValue
	}

	fmt.Printf("Condition OK -> %s = %d\n", op.register, registers[op.register])

	if registers[op.register] > allTimeMax {
		allTimeMax = registers[op.register]
	}
	return allTimeMax
}

func readAllOperations() []Operation {
	f, err := os.Open("input")
	if err != nil {
		panic(err)
	}
	defer f.Close()

	var operations []Operation

	s := bufio.NewScanner(f)
	s.Split(bufio.ScanLines)
	for s.Scan() {
		line := s.Text()
		operation := parseLine(line)
		operations = append(operations, operation)
	}

	return operations
}

func main() {
	operations := readAllOperations()

	// Initialize registers
	registers := make(map[string]int)
	for _, op := range operations {
		if _, ok := registers[op.register]; !ok {
			registers[op.register] = 0
		}
	}

	// Run all operations
	allTimeMax := math.MinInt32
	for _, op := range operations[0:10] {
		allTimeMax = op.run(registers, allTimeMax)
	}

	// Display the largest value in the registers
	max := 0
	for _, value := range registers {
		if value > max {
			max = value
		}
	}
	fmt.Println("Max: ", max)
	fmt.Println("All time max: ", allTimeMax)
}
